File: 3569.maximize-count-of-distinct-primes-after-split.py
#
# @lc app=leetcode id=3569 lang=python3
#
# [3569] Maximize Count of Distinct Primes After Split
#

# @lc code=start
class Solution:
    def maximumCount(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        # Precompute all prime numbers up to 10^5 using Sieve of Eratosthenes
        max_val = 10**5
        is_prime = [True] * (max_val + 1)
        p = 2
        while p * p <= max_val:
            if is_prime[p]:
                for i in range(p * p, max_val + 1, p):
                    is_prime[i] = False
            p += 1
        primes_set = {i for i in range(2, max_val + 1) if is_prime[i]}
        
        results = []
        for idx, val in queries:
            nums[idx] = val # Update nums as per query
            n = len(nums)
            max_distinct_primes = 0
            
            # Calculate prime counts for each possible split point k.
            prefix_primes = set() # Primes in prefix part nums[0..k-1]
            suffix_primes = set(nums) & primes_set # Primes in suffix nums[k..] initially includes all nums elements. 
            
            for k in range(1, n): # Iterate over possible split points k (1 <= k < n) 
                prefix_primes.add(nums[k-1]) # Add element at k-1 to prefix set if it is a prime. 
                suffix_primes.discard(nums[k-1]) # Remove from suffix set as it moves into prefix. 
                max_distinct_primes = max(max_distinct_primes, len(prefix_primes & primes_set) + len(suffix_primes & primes_set)) 
                
            results.append(max_distinct_primes) # Store result after processing each query. 
        return results # Return results list with maximum distinct prime counts after each query update. 
# @lc code=end

File: 3569.maximize-count-of-distinct-primes-after-split.go
#
# @lc app=leetcode id=3569 lang=golang
#
# [3569] Maximize Count of Distinct Primes After Split
#
# @lc code=start
func maximumCount(nums []int, queries [][]int) []int {
    maxVal := 100005
    // Sieve of Eratosthenes to determine primes up to maxVal
    isPrime := make([]bool, maxVal)
    for i := range isPrime {
        isPrime[i] = true
    }
    isPrime[0], isPrime[1] = false, false
    for i := 2; i*i < maxVal; i++ {
        if isPrime[i] {
            for j := i * i; j < maxVal; j += i {
                isPrime[j] = false
            }
        }
    }

    n := len(nums)
    ans := make([]int, 0, len(queries))
    for _, q := range queries {
        idx, val := q[0], q[1]
        nums[idx] = val
        // Build prefix and suffix prime distinct count
        prefixSet := make(map[int]struct{})
        suffixSet := make(map[int]struct{})
        prefixCount := make([]int, n+1)
        suffixCount := make([]int, n+1)
        for i := 0; i < n; i++ {
            if nums[i] < maxVal && isPrime[nums[i]] {
                prefixSet[nums[i]] = struct{}{}
            }
            prefixCount[i+1] = len(prefixSet)
        }
        for i := n - 1; i >= 0; i-- {
            if nums[i] < maxVal && isPrime[nums[i]] {
                suffixSet[nums[i]] = struct{}{}
            }
            suffixCount[i] = len(suffixSet)
        }
        maxDistinct := 0
        for k := 1; k < n; k++ {
            total := prefixCount[k] + suffixCount[k]
            if total > maxDistinct {
                maxDistinct = total
            }
        }
        ans = append(ans, maxDistinct)
    }
    return ans
}
# @lc code=end